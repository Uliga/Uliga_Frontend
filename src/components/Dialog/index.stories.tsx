import React from "react";
import { withKnobs, text } from "@storybook/addon-knobs";
import Dialog from "./index";

export default {
  title: "components|basic/Dialog",
  component: Dialog,
  parameters: {
    docs: {
      inlineStories: false,
    },
  },
  decorators: [withKnobs], // 애드온 적용
};

export const dialog = () => {
  const description = text(
    "description",
    "가계부가 성공적으로 만들어졌습니다. \n 당신의 합리적인 소비생활을 응원합니다! 🙋‍♀️",
  );

  return (
    <Dialog
      size={37}
      title="가계부 생성 완료"
      description={description}
      visible
    />
  );
};

export const cancellable = () => {
  return (
    <Dialog
      size={37}
      title="윤채현님의 가계부 삭제"
      description="정말 윤채현님의 가계부를 삭제하시겠어요?"
      cancellable
      visible
    />
  );
};
export function ControlTest(args: any) {
  return <Dialog {...args} />;
}
ControlTest.args = {
  title: "김동준님의 가계부 삭제",
  description: "정말 김동준님의 가계부를 삭제하시겠어요?",
  cancellable: true,
  visible: true,
};
