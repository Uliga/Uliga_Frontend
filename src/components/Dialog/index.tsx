import React from "react";
import styled from "styled-components";
import Button from "../Button";
import COLORS from "../../constants/color";

export type DialogProps = {
  /** 다이얼로그 너비 */
  size?: number;
  /** 다이얼로그 너비 */
  visible?: boolean;
  /** 다이얼로그 제목 */
  title?: string;
  /** 다이얼로그 내용 */
  description?: string;
  children?: React.ReactNode;
  /** 버튼 유무(확인,취소 둘다) */
  hideButtons?: boolean;
  /** 취소 버튼 유무 */
  cancellable?: boolean;
  /** 취소 버튼 눌렸을때 나올 함수 */
  onCancel?: () => void;
  /** 확인 버튼 눌렸을때 나올 함수 */
  onConfirm?: () => void;
};

const DarkLayer = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  z-index: 999;
  background-color: rgba(0, 0, 0, 0.75);
  //backdrop-filter: blur(1px);
`;

const WhiteBoxWrapper = styled.div`
  position: fixed;
  display: flex;
  align-items: center;
  justify-content: center;
  right: 0;
  top: 0;
  width: 100%;
  height: 100%;
  z-index: 1000;
`;

const ButtonWrapper = styled.div`
  position: absolute;
  bottom: 3rem;
  right: 3rem;
  display: flex;
  z-index: 15;
  gap: 0.8rem;
  margin-left: 17rem;
`;

const WhiteBox = styled.div<{
  size: number;
}>`
  ${({ size }) => `
    width: ${size}rem;
`}

  box-sizing: border-box;
  border-radius: 1rem;
  background: white;
  height: 27rem;
  box-shadow: 0 4px 8px 8px rgba(0, 0, 0, 0.05);
  padding-left: 3rem;
  padding-top: 3.5rem;
  position: relative;
  h2 {
    font-size: 2.1rem;
    font-weight: 700;
    margin-bottom: 2.25rem;
    color: ${COLORS.GREY[600]};
  }

  h4 {
    font-weight: 300;
    margin-bottom: 6.5rem;
    color: ${COLORS.GREY[400]};
    white-space: pre-line;
    width: 31rem;
  }
`;
const StyledConfirmButton = styled(Button)`
  font-size: 1.4rem;
  padding: 1.5rem 0;
  font-weight: 500;
`;
const StyledCancelButton = styled(Button)`
  font-size: 1.4rem;
  padding: 1.5rem 0;
`;

function Dialog({
  size = 45,
  visible = false,
  title,
  description,
  hideButtons = false,
  cancellable = false,
  children,
  onCancel,
  onConfirm,
}: DialogProps) {
  return (
    <>
      {visible && <DarkLayer />}

      {visible && (
        <WhiteBoxWrapper>
          <WhiteBox size={size}>
            {title && <h2>{title}</h2>}
            {description && <h4>{description}</h4>}
            {children}
            {!hideButtons && (
              <ButtonWrapper>
                {cancellable && (
                  <StyledCancelButton
                    theme="unfocus"
                    title="취소"
                    width="8rem"
                    onClick={onCancel}
                  />
                )}
                <StyledConfirmButton
                  theme="primary"
                  title="확인"
                  width={cancellable ? "8rem" : "12rem"}
                  onClick={onConfirm}
                />
              </ButtonWrapper>
            )}
          </WhiteBox>
        </WhiteBoxWrapper>
      )}
    </>
  );
}

export default Dialog;
