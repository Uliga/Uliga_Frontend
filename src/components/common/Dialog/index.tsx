import React from "react";
import styled from "styled-components";
import Button from "../Button";
import COLORS from "../../../constants/color";

export type DialogProps = {
  size: number;
  visible?: boolean;
  title?: string;
  description?: string;
  children?: React.ReactNode;
  hideButtons?: boolean;
  cancellable?: boolean;
  onCancel?: () => void;
  onConfirm?: () => void;
};

const DarkLayer = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  z-index: 10;
  background: rgba(0, 0, 0, 0);
`;

const WhiteBoxWrapper = styled.div`
  position: fixed;
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 100%;
  z-index: 15;
`;

const ButtonWrapper = styled.div`
  position: fixed;
  display: flex;
  z-index: 15;
  gap: 0.5rem;
  margin-left: 17rem;
`;

const WhiteBox = styled.div<{
  size: number;
}>`
  ${({ size }) => `
    width: ${size}rem;
`}
  box-sizing: border-box;
  border-radius: 36px;
  background: white;
  height: 22rem;
  box-shadow: 0 4px 8px 8px rgba(0, 0, 0, 0.05);
  padding-left: 2.9rem;

  h3 {
    font-size: 2rem;
    margin-top: 3rem;
    margin-bottom: 2.25rem;
    color: ${COLORS.BLACK};
  }

  p {
    font-size: 1.125rem;
    margin-bottom: 6.5rem;
    color: ${COLORS.GREY[500]};
    white-space: pre-line;
  }
`;
const StyledConfirmButton = styled(Button)`
  border-radius: 15px;
  padding: 1rem 0 1rem 0;
`;
const StyledCancelButton = styled(Button)`
  border-radius: 15px;
  padding: 1rem 0 1rem 0;
`;
function Dialog({
  size = 37,
  visible = true,
  title,
  description,
  hideButtons = false,
  cancellable = true,
  children,
  onCancel,
  onConfirm,
}: DialogProps) {
  return (
    <>
      {visible && <DarkLayer />}

      {visible && (
        <WhiteBoxWrapper>
          <WhiteBox size={size}>
            {title && <h3>{title}</h3>}
            {description && <p>{description}</p>}
            {children}
            {!hideButtons && (
              <ButtonWrapper>
                {cancellable && (
                  <StyledCancelButton
                    theme="unfocus"
                    size="small"
                    title="취소"
                    width="6.57rem"
                    onClick={onCancel}
                  />
                )}
                <StyledConfirmButton
                  theme="primary"
                  size="small"
                  title="확인"
                  width={cancellable ? "7rem" : "4.57rem"}
                  onClick={onConfirm}
                />
              </ButtonWrapper>
            )}
          </WhiteBox>
        </WhiteBoxWrapper>
      )}
    </>
  );
}

export default Dialog;
