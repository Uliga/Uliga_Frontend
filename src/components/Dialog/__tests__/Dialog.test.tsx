import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Dialog from "../index"; // 다이얼로그 컴포넌트의 실제 파일 경로로 변경해야 합니다.

describe("<Dialog />", () => {
  test("확인 버튼이 클릭되었을 때 동작을 확인합니다", () => {
    let confirmClicked = false; // 확인 버튼 클릭 여부를 저장할 변수
    const handleConfirm = () => {
      confirmClicked = true;
    };

    const { getByText } = render(
      <Dialog
        visible
        title="제목"
        description="설명"
        onConfirm={handleConfirm}
      />,
    );

    const confirmButton = getByText("확인");
    fireEvent.click(confirmButton);

    expect(confirmClicked).toBe(true);
  });
  test("확인 버튼이 클릭되었을 때 onConfirm 이벤트가 호출되는지 확인합니다", () => {
    const onConfirm = jest.fn();

    const { getByText } = render(
      <Dialog visible title="제목" description="설명" onConfirm={onConfirm} />,
    );

    const confirmButton = getByText("확인");
    fireEvent.click(confirmButton);

    expect(onConfirm).toHaveBeenCalled();
  });
  test("취소 버튼이 클릭되었을 때 onCancel 이벤트가 호출되는지 확인합니다", () => {
    const onCancel = jest.fn();

    const { getByText } = render(
      <Dialog
        visible
        title="제목"
        description="설명"
        cancellable
        onCancel={onCancel}
      />,
    );

    const cancelButton = getByText("취소");
    fireEvent.click(cancelButton);

    expect(onCancel).toHaveBeenCalled();
  });

  test("다이얼로그의 Title과 Description이 스냅샷과 일치해야 합니다", () => {
    const { container } = render(
      <Dialog visible title="Test Title" description="Test Description" />,
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
