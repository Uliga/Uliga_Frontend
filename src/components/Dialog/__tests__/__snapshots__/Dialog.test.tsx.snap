// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dialog /> 다이얼로그의 Title과 Description이 스냅샷과 일치해야 합니다 1`] = `
<div
  class="sc-gswNZR ebuQKt"
/>
`;
